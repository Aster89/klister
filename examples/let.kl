#lang "n-ary-app.kl"

(import "quasiquote.kl")
(import (shift "n-ary-app.kl" 1))
(import (shift "list.kl" 1))
(import (shift "quasiquote.kl" 1))

(define-macros
  -- (let ((x foo)
  --       (y bar))
  --   body)
  -- =>
  -- ((lambda (x y) body)
  --  foo
  --  bar)
  ((let (lambda (stx)
          (syntax-case stx
            ((list (_ args body))
             (pure (cons-list-syntax
                     `(lambda ,(map car args) ,body)
                     (map cdr args)
                     stx))))))))

-- examples from Racket's documentation of let

-- 'five
(example (let ([x 'five]) x))

-- ('five 'two)
(example (let ([x 'five])
           (let ([x 'two]
                 [y x])
             `(,y ,x))))


(export let)
