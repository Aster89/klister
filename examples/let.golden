'five : Syntax
'(five two) : Syntax
